src/mlpack/methods/ann/activation_functions/lisht_function.hpp

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file lisht_function.hpp
+ * @author Kartik Dutt
+ *
+ * Definition and implementation of the LiSHT function as described by
+ * Swalpa K. Roy, Suvojit Manna, Shiv Ram Dubey and Bidyut B. Chaudhuri.
+ *
+ * For more information, see the following paper.
+ *
+ * @code
+ * @misc{
+ *   author = {Swalpa K. Roy, Suvojit Manna, Shiv R. Dubey and
+ *            Bidyut B. Chaudhuri},
+ *   title = {LiSHT: Non-Parametric Linearly Scaled Hyperbolic Tangent
+ *           Activation Function for Neural Networks},
+ *   year = {2019}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_LISHT_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_LISHT_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <algorithm>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Lisht function, defined by
+ *
+ * @f{eqnarray*}{
+ * f(x) = x * tanh(x)
+ * f'(x) = tanh(x) + x * (1 - tanh^{2}(x))
+ * @f}
+ */
+class LishtFunction
+{
+ public:
+  /**
+   * Computes the Lisht function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x)
+  {
+    return x * std::tanh(x);
+  }
+
+  /**
+   * Computes the Lisht function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template <typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType &x, OutputVecType &y)
+  {
+    y = x % arma::tanh(x);
+  }
+
+  /**
+   * Computes the first derivative of the Lisht function.
+   *
+   * @param y Input data.
+   * @return f'(x)
+   */
+  static double Deriv(const double y)
+  {
+    return (4 * y * std::exp(2 * y) + std::exp(4 * y) - 1) /
```

Comment:
Is this some kind of approximation for the derivative as described in the paper?

Reply:
Hi, no this is the same expression as that in the paper but simplified further in exponentials to make it a bit faster.

Reply:
Thanks for the clarification.

Reply:
Do you think this is faster as using `tanh`? I mean if I would use `tanh` the expression would look at lot simpler, but if that one is faster, I don't mind to use that instead but in this case let's reference the formula from the paper right above.

Reply:
Hi @zoq, I am changing to paper's definition here since I ran a test for both to compare time to get following results on an input of 100000000
Time: 8.66564
Time: 8.68422
Which I think is within in margin of error and the definition in paper looks cleaner. I thought it must be much faster as it was [here.](https://github.com/digantamisra98/Mish/issues/22#issuecomment-573394994)

Reply:
Agreed, the current implementation looks cleaner, I think the compiler already does a pretty good job at optimizing `tanh`.